sources/models/src/modeled_types/shared.rs

```diff
@@ -130,6 +130,102 @@ mod test_single_line_string {
 
 // =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
 
+/// ValidLinuxHostname represents a string that contains a valid Linux hostname as defined by
+/// https://man7.org/linux/man-pages/man7/hostname.7.html.  It stores the original form and makes
+/// it accessible through standard traits.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct ValidLinuxHostname {
+    inner: String,
+}
+
+lazy_static! {
+    // According to the man page above, hostnames must be between 1-253 characters long consisting
+    // of characters [0-9a-z.-].
+    pub(crate) static ref VALID_LINUX_HOSTNAME: Regex = Regex::new(r"^[0-9a-z.-]{1,253}$").unwrap();
+}
+
+impl TryFrom<&str> for ValidLinuxHostname {
+    type Error = error::Error;
+
+    fn try_from(input: &str) -> Result<Self, Self::Error> {
+        ensure!(
+            VALID_LINUX_HOSTNAME.is_match(input),
+            error::InvalidLinuxHostname {
+                input,
+                msg: "must only be [0-9a-z.-], and 1-253 chars long"
+            }
+        );
+
+        // Though the man page doesn't explicitly disallow hostnames that start with dots, dots are
+        // used as separators so starting with a separator would imply an empty domain, which isn't
+        // allowed (must be at least one character).
+        ensure!(
+            !input.starts_with("-") && !input.starts_with("."),
+            error::InvalidLinuxHostname {
+                input,
+                msg: "must not start with '-' or '.'"
+            }
+        );
+
+        // Each segment must be from 1-63 chars long
+        ensure!(
+            input.split(".").all(|x| x.len() >= 1 && x.len() <= 63),
+            error::InvalidLinuxHostname {
+                input,
+                msg: "segment is less than 1 or greater than 63 chars"
+            }
+        );
```

Comment:
It also seems like segments shouldn't start with a hyphen, even if that's not specified in the man page.

Reply:
Good point - I'll fix that.